.github/workflows/ci.yml

```diff
@@ -10,55 +10,25 @@ concurrency:
   group: ${{ github.workflow }}-${{ github.ref }}
   cancel-in-progress: true
 jobs:
-  analyze:
+  tox:
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.10", "3.x"]
     runs-on: ubuntu-latest
     permissions:
       contents: read
     steps:
-    - name: Checkout repository
-      uses: actions/checkout@v3
-
-    - name: Setup virtual environment
+    - uses: actions/checkout@v3
+    - uses: actions/setup-python@v4
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Install dependencies
       run: |
-        pip3 install virtualenv
-        python3 -m venv .venv
-        source .venv/bin/activate
-        pip3 install -r ./requirements.txt
-        pip3 install flake8 pycodestyle isort mypy black
-        mypy --install-types --non-interactive pytest_mh example tests
-
-    - name: flake8
-      run: source .venv/bin/activate && flake8 .
-
-    - name: pycodestyle
-      run: source .venv/bin/activate && pycodestyle .
-
-    - name: isort
-      run: source .venv/bin/activate && isort --check-only .
-
-    - name: black
-      run: source .venv/bin/activate && black --check -
-
-    - name: mypy
-      run: source .venv/bin/activate && mypy pytest_mh example tests
-
-  tests:
-    runs-on: ubuntu-latest
-    permissions:
-      contents: read
-    steps:
-    - name: Checkout repository
-      uses: actions/checkout@v3
-
-    - name: Setup virtual environment
-      run: |
-        pip3 install virtualenv
-        python3 -m venv .venv
-        source .venv/bin/activate
-        pip3 install -r ./requirements.txt
-
-    - name: Run tests
-      run: source .venv/bin/activate && pytest --color=yes -vv
+        python -m pip install --upgrade pip
+        pip install tox tox-gh
+    - name: Run tox
+      run: tox --colored=yes
```

Comment:
@pbrezina FYI it's customary to split the env provisioning and testing into two commands so that it's visible in the CI which step fails. This example shows how to do this: https://github.com/tox-dev/tox-gh#workflow-configuration.

Also, here's a neat trick to re-run only the tests that failed with maximum verbosity, keeping the job status failing: https://github.com/cherrypy/cheroot/blob/0fd16f0/.github/workflows/ci-cd.yml#L880-L895. I thought you might find it interesting.

Reply:
Cool, thank you.